core/shared/src/main/scala/scalaz/zio/Managed.scala

```diff
@@ -11,44 +11,39 @@ package scalaz.zio
  * has been consumed, the resource will not be valid anymore and may fail with
  * some checked error, as per the type of the functions provided by the resource.
  */
-sealed abstract class Managed[+E, +R] extends Serializable { self =>
-  type R0 <: R
+final case class Managed[+E, +R](reserve: IO[E, Managed.Reservation[E, R]]) { self =>
+  import Managed.Reservation
 
-  protected def acquire: IO[E, R0]
-
-  protected def release: R0 => IO[Nothing, _]
-
-  def use[E1 >: E, A](f: R => IO[E1, A]): IO[E1, A]
+  def use[E1 >: E, A](f: R => IO[E1, A]): IO[E1, A] =
+    reserve.flatMap { token =>
+      token.acquire.flatMap(f).ensuring(token.release)
```

Comment:
I think more precisely we need: `reserve.bracket(_.release)(_.acquire.flatMap(f))`.

Because otherwise, `reserve` may succeed but the returned `IO` may be interrupted before it has a chance to `flatMap` over the next action, which installs the `token.release` finalizer.

Reply:
Oops, thank you, missed this one.